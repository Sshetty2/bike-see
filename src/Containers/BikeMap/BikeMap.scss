.map-container {
  margin: 0 auto;
  width: 100%;
  height: 100vh;
}

#map {
  display: block;
  margin: 0;
  border: none;
  width: 100%;
  height: 100%;
  z-index: 9;
}
