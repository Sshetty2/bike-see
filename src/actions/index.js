export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading,
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error,
});

export const setCities = (cities) => ({
  type: 'SET_CITIES',
  cities,
});

export const setCurrentCity = (city) => ({
  type: 'SET_CURRENT_CITY',
  city,
});

export const setStations = (stations) => ({
  type: 'SET_STATIONS',
  stations,
});

export const toggleFavorite = (favorite) => ({
  type: 'TOGGLE_FAVORITE',
  favorite,
});