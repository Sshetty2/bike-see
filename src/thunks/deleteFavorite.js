import { removeFavorite, setLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const deleteFavorite = (station_id, user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      await fetchData(`https://bike-see.herokuapp.com/api/v1/users/${user_id}/favorites/${station_id}`, 'DELETE');
      dispatch(removeFavorite(station_id));
    } catch (error) {
      dispatch(setError('Error deleting favorite.'));
    }
    dispatch(setLoading(false));
  }
}