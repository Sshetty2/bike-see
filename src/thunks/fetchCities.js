import { setLoading, setError, setCities } from '../actions';
import { fetchData } from '../utils/api';

export const fetchCities = () => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const results = await fetchData('https://api.citybik.es/v2/networks', 'GET');
      dispatch(setCities(results.networks));
    } catch (error) {
      dispatch(setError('Error fetching cities.'));
    }
    dispatch(setLoading(false));
  }
}