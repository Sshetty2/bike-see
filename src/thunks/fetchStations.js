import { setLoading, setError, setStations } from '../actions';
import { fetchData } from '../utils/api';

export const fetchStations = (path) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const results = await fetchData(`http://api.citybik.es/v2/networks/${path}`);
      dispatch(setStations(results.network.stations));
    } catch (error) {
      dispatch(setError(error));
    }
    dispatch(setLoading(false));
  }
}