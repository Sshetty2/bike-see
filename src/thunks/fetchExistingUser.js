import { setCurrentUser, setLoading, setError } from '../actions';
import { fetchFavorites } from '../thunks/fetchFavorites';
import { fetchData } from '../utils/api';

export const fetchExistingUser = (id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedUser = await fetchData(`/api/v1/users/${id}`, 'GET');
      dispatch(setCurrentUser(fetchedUser));
      dispatch(fetchFavorites(fetchedUser.id));
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}