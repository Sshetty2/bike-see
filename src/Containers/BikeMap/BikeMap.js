import React, { Component } from 'react';
import { Map, Marker, Tooltip, TileLayer, withLeaflet } from 'react-leaflet';
import { ReactLeafletSearch } from 'react-leaflet-search';
import L from 'leaflet';
import { connect } from 'react-redux';
import { fetchStations } from '../../thunks/fetchStations';
import { postFavorite } from '../../thunks/postFavorite';
import { deleteFavorite } from '../../thunks/deleteFavorite';
import PropTypes from 'prop-types';
import getDistance from 'geodist';

const MapSearch = withLeaflet(ReactLeafletSearch);

export class BikeMap extends Component {
  constructor(props) {
    super(props);
    this.state = {
      lat: 0,
      lon: 0,
      loading: true,
    }
  }

  getLocation = () => {
    this.setState({ lat: 0, lon: 0 });
    navigator.geolocation.getCurrentPosition((position) => {
      this.setState({
        lat: position.coords.latitude,
        lon: position.coords.longitude
      });
    });
  }

  showCurrentLocation = () => {
    const { lat, lon } = this.state;
    return (
      <Marker
        position={[lat, lon]}
        key={'geoloc'}
        id={'geoloc'}>
        <Tooltip className='tooltip'>{<h4>You Are Here</h4>}</Tooltip>
      </Marker>
    )
  }

  getStations = (e) => {
    const { fetchStations, history } = this.props;
    fetchStations(e.target.options.id);
    localStorage.setItem('current-city', JSON.stringify(e.target.options.id));
    history.replace('/stations');
  }

  toggleFavorite = (e) => {
    const { id } = e.target.options;
    const { deleteFavorite, postFavorite, user, favorites } = this.props;
    favorites.includes(e.target.options.id) ? 
      deleteFavorite(id, user.id) :
      postFavorite(id, user.id);
    // this.setLocalStorage();
  }

  setLocalStorage = () => {
    localStorage.setItem('bike-stops', JSON.stringify(this.props.favorites));
  }

  createStationMarkers = (data, icon) => {
    return data.map(marker => {
      const { name, latitude, longitude, free_bikes, empty_slots, timestamp, id } = marker;
      const { favorites } = this.props;

      let newIcon;
      let buttonText;
      if (favorites.includes(id)) {
        newIcon = new L.icon({ ...icon, iconUrl: require('../../images/bike-purple.png') }); 
        buttonText = 'Click to remove from stops';
      } else {
        newIcon = new L.icon({ ...icon, iconUrl: require('../../images/bike.png') });
        buttonText = 'Click to add to stops';
      }
      
      const distance = getDistance({ lat: this.state.lat, lon: this.state.lon }, { lat: latitude, lon: longitude }, { exact: true, unit: 'mi' });
      const temp = new Date(timestamp);
      const date = temp.toDateString();
      const time = temp.toTimeString().substring(0,5);
      return (
        <Marker
          onClick={this.toggleFavorite}
          position={[latitude, longitude]}
          icon={newIcon}
          key={id}
          id={id}>
          <Tooltip className='tooltip'>{
            <div>
              <h4>{name}</h4>
              <p>Distance away: {distance.toFixed(1)} mi</p>
              <p>Empty slots: {empty_slots}</p>
              <p>Free bikes: {free_bikes}</p>
              <p>Updated: {date}, {time}</p>
              <p className='click-text'>{buttonText}</p>
            </div>
          }</Tooltip>
        </Marker>
      )
    });
  }

  createCityMarkers = (data, icon) => {
    return data.map(marker => {
      const { name, id } = marker;
      const { city, country, latitude, longitude } = marker.location;
      return (
        <Marker
          onClick={this.getStations}
          position={[latitude, longitude]}
          icon={icon}
          key={id}
          id={id}>
          <Tooltip className='tooltip'>{
            <div>
              <h4>{name}</h4>
              <h4>{city}, {country}</h4>
              <p className='click-text'>Click to view stations</p>
            </div>}
          </Tooltip>
        </Marker>
      )
    });
  }

  showMarkers = () => {
    const { pathname } = this.props.location;
    const { stations, cities, favorites } = this.props;
    let icon = {
      iconSize: [25, 25],
      iconAnchor: [12, 41],
      popupAnchor: [1, -25],
    };
    let data;

    switch (pathname) {
      case '/my-stops':
        data = stations.filter(station => favorites.includes(station.id));
        return this.createStationMarkers(data, icon)
      case '/stations':
        data = stations;
        return this.createStationMarkers(data, icon)
      default:
        const newIcon = new L.icon({
          iconSize: [25, 41],
          iconAnchor: [12, 41],
          popupAnchor: [1, -34],
          shadowSize: [41, 41],
          shadowUrl: require('../../images/marker-shadow.png'),
          iconUrl: require('../../images/marker-icon-violet.png')
        });
        data = cities;
        return this.createCityMarkers(data, newIcon)
    }
  }

  componentDidMount() {
    this.getLocation();
    setTimeout(() => {
      this.setState({ loading: false })
    }, 100);
  }

  render() {
    const { lat, lon, loading } = this.state;
    return (
      <div className='map-container'>
        {!loading && <i onClick={this.getLocation} className="fas fa-location-arrow"></i>}
        {!loading && 
          <Map
            id='map'
            minZoom='3'
            maxZoom='19'
            center={[lat, lon]}
            zoom='13'>
            {this.showCurrentLocation()}
            {this.showMarkers()}
            <MapSearch 
              position='topright'
              inputPlaceholder="Search by location"
              showMarker={false}
              showPopup={true}
              openSearchOnLoad={true}
            />
            <TileLayer
              url='https://{s}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}{r}.png'
              attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a> | <a href="http://api.citybik.es/v2/">CityBikes API</a>'
              subdomains='abcd'
              minZoom='3'
              maxZoom='19'
              ext='png'
            />
          </Map>
        }
      </div>  
    )
  }
}

export const mapStateToProps = (state) => ({
  cities: state.cities,
  stations: state.stations,
  favorites: state.favorites,
  user: state.user,
});

export const mapDispatchToProps = (dispatch) => ({
  fetchStations: (id) => dispatch(fetchStations(id)),
  deleteFavorite: (station, user) => dispatch(deleteFavorite(station, user)),
  postFavorite: (station, user) => dispatch(postFavorite(station, user))
});

BikeMap.propTypes = {
  cities: PropTypes.array,
  stations: PropTypes.array,
  favorites: PropTypes.array,
  fetchStations: PropTypes.func.isRequired,
}

BikeMap.defaultProps = {
  cities: [],
  stations: [],
  favorites: []
}

export default connect(mapStateToProps, mapDispatchToProps)(BikeMap);