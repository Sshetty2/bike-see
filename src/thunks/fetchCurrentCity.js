import { setLoading, setError, setCurrentCity } from '../actions';
import { fetchStations } from '../thunks/fetchStations';
import { fetchData } from '../utils/api';

export const fetchCurrentCity = (user_id) => {
  return async (dispatch) => {
    try {
      dispatch(setLoading(true));
      const fetchedCity = await fetchData(`http://localhost:3001/api/v1/users/${user_id}/city`, 'GET');

      if (fetchedCity !== '') {
        dispatch(setCurrentCity(fetchedCity.city));
        dispatch(fetchStations(user_id, fetchedCity.city));
      }

    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  }
}