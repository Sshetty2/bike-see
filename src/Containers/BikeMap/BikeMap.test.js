import React from 'react';
import { BikeMap, mapStateToProps, mapDispatchToProps } from './BikeMap';
import { shallow } from 'enzyme';
import { fetchStations } from '../../thunks/fetchStations';
import { toggleFavorite } from '../../actions';
jest.mock('../../thunks/fetchStations');

describe('BikeMap', () => {
  let wrapper;
  let fetchStationsMock;
  let toggleFavoriteMock;
  let mockCities;
  let mockStations;
  let mockFavorites;
  let mockEvent;
  let mockHistory;

  beforeEach(() => {
    mockCities = [
      { name: 'Denver', stations: 26, id: 'denver' },
      { name: 'Capital Bikeshare', stations: 58, id: 'capital-bikeshare' }
    ];
    mockStations = [
      { name: '1st & Broadway', bikes_available: 5, id: '1' },
      { name: '18th & Blake', bikes_available: 8, id: '2' }
    ];
    mockFavorites = ['1'];
    fetchStationsMock = jest.fn();
    toggleFavoriteMock = jest.fn();
    mockEvent = { target: { options: {id: '1'} } };
    mockHistory = { replace: jest.fn() };
    global.navigator.geolocation = {getCurrentPosition: jest.fn()};
    wrapper = shallow(
      <BikeMap
        cities={mockCities}
        stations={mockStations}
        favorites={mockFavorites}
        fetchStations={fetchStationsMock}
        toggleFavorite={toggleFavoriteMock}
        history={mockHistory}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('getLocation', () => {
    it('should call navigator.geolocation.getCurrentPosition', () => {
      wrapper.instance().getLocation();
      expect(global.navigator.geolocation.getCurrentPosition).toHaveBeenCalled();
    });
  });

  describe('getStations', () => {
    it('should call fetchStations with the correct params', () => {
      wrapper.instance().getStations(mockEvent);
      expect(fetchStationsMock).toHaveBeenCalledWith('1');
    });

    it('should update the path', () => {
      wrapper.instance().getStations(mockEvent);
      expect(wrapper.instance().props.history.replace).toHaveBeenCalledWith('/stations');
    });
  });

  describe('toggleFavorite', () => {
    it('should call toggleFavorite with the correct params', () => {
      wrapper.instance().toggleFavorite(mockEvent);
      expect(wrapper.instance().props.toggleFavorite).toHaveBeenCalledWith('1');
    });

    it('should call setLocalStorage', () => {
      wrapper.instance().setLocalStorage = jest.fn();
      wrapper.instance().toggleFavorite(mockEvent);
      expect(wrapper.instance().setLocalStorage).toHaveBeenCalled();
    });
  });

  describe('createStationMarkers', () => {
    it('', () => {

    });
  });

  describe('createCityMarkers', () => {
    it('', () => {

    });
  });

  describe('showMarkers', () => {
    it('', () => {

    });
  });

  describe('componentDidMount', () => {
    it('should call getLocation', () => {
      wrapper.instance().getLocation = jest.fn();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().getLocation).toHaveBeenCalled();
    });

    it('should set loading in state to false', () => {
      wrapper.instance().getLocation = jest.fn();
      wrapper.instance().componentDidMount();
      expect(wrapper.state('loading')).toEqual(false);
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        cities: mockCities,
        stations: mockStations,
        favorites: mockFavorites,
        otherState: false,
      };
      const expected = {
        cities: mockCities,
        stations: mockStations,
        favorites: mockFavorites
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when fetchStations is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchStations();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchStations();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch when toggleFavorite is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = toggleFavorite();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.toggleFavorite();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});