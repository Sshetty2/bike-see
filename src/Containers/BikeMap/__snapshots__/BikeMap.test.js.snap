// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BikeMap should match the correct snapshot 1`] = `
<div
  className="map-container"
/>
`;
